iterators.tex

```diff
@@ -6327,19 +6327,26 @@
 
 \begin{itemdecl}
 template <class T>
-using IteratorType = decltype(begin(declval<T>()));
+  @\newtxt{requires requires (T\& t) \{}@
+    @\newtxt{\{ begin(t) \} -> Iterator; // not required to be equality preserving}@
+  @\newtxt{\}}@
+using IteratorType = decltype(begin(declval<T@\newtxt{\&}@>()));
 
 template <class T>
-using SentinelType = decltype(end(declval<T>()));
+  @\newtxt{requires requires (T\& t) \{}@
+    @\newtxt{typename IteratorType<T>;}@
+    @\newtxt{\{ end(t) \} -> Sentinel<IteratorType<T>{}>; // not required to be equality preserving}@
+  @\newtxt{\}}@
+using SentinelType = decltype(end(declval<T@\newtxt{\&}@>()));
```

Comment:
These definitions have the curious property that I can call `begin` on an object that doesn't have an `end`, but I can't call `end` on an object that doesn't have a `begin`. What's the rationale?